### PR TITLE
feat: implement websocket in backoffice-v2 and workflows-service

### PR DESCRIPTION
**Pull request by TzlilSwimmer123**
installed `react-use-websocket` package for implementing websockets in backoffice-v2, connect to websocket and wait for update messages instead of polling. if disconnected start polling again. workflows-service send api calls to websocket server to send update messages to all clients

BREAKING CHANGE: backoffice-v2 connects to websocket and if fails keep polling in order to present updated data

feat #534

### Description
Enhance the workflow service and backoffice application by replacing the current polling strategy with real-time updates through websockets.

### Related issues
 https://github.com/ballerine-io/ballerine/issues/534

### Breaking changes
 backoffice-v2 connects to websocket and if fails keep polling in order to present updated data

### How these changes were tested
 * At the moments all the tests are manual using postman

### Examples and references
 * Links, screenshots, and other resources related to this change.

### Checklist
- [X] I have read the [contribution guidelines](CONTRIBUTING.md) of this project
- [X] I have read the [style guidelines](STYLE_GUIDE.md) of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [?] My changes generate no new warnings and errors
- [] New and existing tests pass locally with my changes

### CODE REVIEW

File: apps/backoffice-v2/src/common/env/schema.ts [hunk 1]
```diff
@@ -17,4 +17,5 @@ export const EnvSchema = z.object({
     .or(z.literal(false))
     .catch(undefined),
   VITE_IMAGE_LOGO_URL: z.string().optional(),
+  WEBHOOK_URL: z.string().url().default('ws://localhost:3500'),
```
Comment (TzlilSwimmer123) [inline apps/backoffice-v2/src/common/env/schema.ts:20]: there is an `.env` that I missed which make the app get `❌ Invalid environment variables`

File: apps/backoffice-v2/src/common/env/schema.ts [hunk 2]
```diff
@@ -17,4 +17,5 @@ export const EnvSchema = z.object({
     .or(z.literal(false))
     .catch(undefined),
   VITE_IMAGE_LOGO_URL: z.string().optional(),
+  WEBSOCKET_URL: z.string().url().default('ws://localhost:3500'),
```
Comment (TzlilSwimmer123) [inline apps/backoffice-v2/src/common/env/schema.ts:20]: I added this variable in the 2 `.env` files but it still throw `❌ Invalid environment variables` so until I will figure it out

Comment (Omri-Levy) [reply apps/backoffice-v2/src/common/env/schema.ts:20]: Hey @TzlilSwimmer123 , there's a chance `zod` doesn't consider `ws://` a valid URL. We can use a regex match here or a custom validation doing a string comparison i.e. `url.startsWith('ws://')` and check if anything past `ws://` is a valid URL. Great job on this PR!

Comment (TzlilSwimmer123) [reply apps/backoffice-v2/src/common/env/schema.ts:20]: @Omri-Levy thanks:)

Comment (TzlilSwimmer123) [reply apps/backoffice-v2/src/common/env/schema.ts:20]: @Omri-Levy I checked it and have 2 comments:
1. if I put `.default('sdfsdf')` - it fails. It means zod consider it as a valid url
2. I printed the problematic env and this is it:
```
{
[1]     "VITE_API_URL": "http://localhost:3000/api/v1/internal",
[1]     "VITE_API_KEY": "demo",
[1]     "VITE_AUTH_ENABLED": "true",
[1]     "VITE_MOCK_SERVER": "false",
[1]     "VITE_POLLING_INTERVAL": "3",
[1]     "VITE_ASSIGNMENT_POLLING_INTERVAL": "5",
[1]     "VITE_EXAMPLE_TYPE": "kyb",
[1]     "BASE_URL": "/",
[1]     "MODE": "development",
[1]     "DEV": true,
[1]     "PROD": false,
[1]     "SSR": false
[1]   }
```
I didn't find where it comes from :/

Comment (TzlilSwimmer123) [reply apps/backoffice-v2/src/common/env/schema.ts:20]: anyway, I removed it and not it's ok 🦤

File: apps/backoffice-v2/src/common/env/schema.ts (hunk 1 above)
Comment (Omri-Levy) [reply apps/backoffice-v2/src/common/env/schema.ts]: Hey @TzlilSwimmer123, can we resolve the comments and fix the merge conflicts?

File: services/workflows-service/src/filter/filter.controller.external.ts [hunk 3]
```diff
@@ -59,11 +62,15 @@ export class FilterControllerExternal {
...
+    const createdFilter = await this.service.create({
       data: {
         ...data,
         query: data?.query as InputJsonValue,
       },
     });
+    const websocketServerNotifyUri = `${env.WEBSOCKET_URL}/notify?type=filters`;
+    // todo is it important to await this?
+    await axios.post(websocketServerNotifyUri);
+    return createdFilter;
```
Comment (TzlilSwimmer123) [inline services/workflows-service/src/filter/filter.controller.external.ts:74]: @Omri-Levy what do you think?

Comment (TzlilSwimmer123) [inline services/workflows-service/src/filter/filter.controller.external.ts:74]: @Omri-Levy what do you think?

File: apps/backoffice-v2/src/common/env/schema.ts (hunk 1 above)
Comment (TzlilSwimmer123) [reply apps/backoffice-v2/src/common/env/schema.ts]: on it, left a question

File: services/workflows-service/src/filter/filter.controller.external.ts (hunk 3 above)
Comment (Omri-Levy) [reply services/workflows-service/src/filter/filter.controller.external.ts:74]: Removing the `await` does seem to make sense here. If you do, please confirm how a rejected promise behaves.

File: services/workflows-service/src/workflow/workflow.controller.internal.unit.test.ts [hunk 4]
```diff
@@ -121,6 +122,7 @@ describe('WorkflowControllerInternal', () => {
             numb: 'context',
           },
         };
+        axios.post = jest.fn().mockResolvedValue({});
```
Comment (Omri-Levy) [inline services/workflows-service/src/workflow/workflow.controller.internal.unit.test.ts:125]: Check out `nock`, `,msw`, and [the following -](https://kentcdodds.com/blog/stop-mocking-fetch) (no need for changes)

File: services/websocket-service/src/notifications/notify.controller.ts [hunk 5]
```diff
@@ -1,13 +1,28 @@
-import { Controller, Post } from '@nestjs/common';
+import { Controller, Post, Query } from '@nestjs/common';
 import { AppGateway } from '@/app.gateway';
+import { IsDefined, IsEnum, IsNotEmpty, IsString } from 'class-validator';
+
+enum NotificationType {
```
Comment (Omri-Levy) [inline services/websocket-service/src/notifications/notify.controller.ts:5]: Hey @TzlilSwimmer123 , prefer `as const` objects to enums. See https://www.youtube.com/watch?v=jjMbPt_H3RQ

**Pull request closed without merging**